src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,241 @@
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
 
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
+        cartEntryRepository.delete(cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().get());
 
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
-        }
+        cart.setEntries(cart.getEntries().stream().filter(entry -> !productName.equals(entry.getProductName())).collect(Collectors.toList()));
+        cart.setTotal(cart.getTotal() - cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).count() * productDTO.getPrice());
 
-        if (Math.toIntExact(productsInCart) < quantity) {
-            throw new InvalidInputDataException("You cannot remove more than " + productsInCart + " " + productName
-                    + " Products from the shopping cart!");
-        } else {
-            double productTotal = 0;
-
-            for (int i = 0; i < quantity; i++) {
-                productList.remove(product);
-                productTotal += product.getProductPrice();
-            }
-
-            cart.setUser(user);
-
-            cart.setProductList(productList);
-            cart.setLastUpdatedOn(LocalDateTime.now());
-            cart.setTotal(cart.getTotal() - productTotal);
-
-            cartRepository.save(cart);
+        cartRepository.save(cart);
 
-            return CartMapper.cartEntityToDto(cart);
-        }
+        return cartEntityToDto(cartRepository.findByUserId(user.getId()));
     }
 
-    public void deleteCartByEmail(String email) throws  EntityNotFoundException {
-
-        User user;
+    public void deleteCartByEmail(String email) throws EntityNotFoundException {
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
 
-        //TODO you might benefit from hibernate caching but you might also do the same DB query twice here(once when you want to see it's not null and 2nd time when you use it)
-        if (cartRepository.findByUser(user) != null) {
-            Cart cart = cartRepository.findByUser(user);
+        if (cartRepository.findByUserId(user.getId()) != null) {
```

Comment:
why do you call findByUserId twice?
why don't you do something like
`Cart cart = cartRepository.findByUserId(user.getId);
if(cart==null){
//throw exception
}`
//logic related to cart found

Reply:
Done!